File: sections/33_inclusion_exclusion/problems/c_bounded_candies/solution.cpp
#include <bits/stdc++.h>
using namespace std;

const long long MOD = 1000000007LL;

long long mod_pow(long long a, long long b) {
    long long result = 1;
    while (b > 0) {
        if (b & 1LL) result = result * a % MOD;
        a = a * a % MOD;
        b >>= 1LL;
    }
    return result;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int q;
    if (!(cin >> q)) return 0;
    vector<array<int, 3>> queries(q);
    int limit = 0;
    for (auto &cur : queries) {
        cin >> cur[0] >> cur[1] >> cur[2];
        limit = max(limit, cur[0] + cur[1]);
    }
    vector<long long> fact(limit + 1, 1), invfact(limit + 1, 1);
    for (int i = 1; i <= limit; i++) fact[i] = fact[i - 1] * i % MOD;
    invfact[limit] = mod_pow(fact[limit], MOD - 2);
    for (int i = limit; i >= 1; i--) invfact[i - 1] = invfact[i] * i % MOD;
    auto C = [&](int n, int r) -> long long {
        if (r < 0 || r > n) return 0;
        return fact[n] * invfact[r] % MOD * invfact[n - r] % MOD;
    };
    for (auto [n, s, b] : queries) {
        long long ans = 0;
        int step = b + 1;
        for (int bad = 0; bad <= n && bad * step <= s; bad++) {
            int remaining = s - bad * step;
            long long term = C(n, bad) * C(remaining + n - 1, n - 1) % MOD;
            if (bad % 2) ans = (ans - term + MOD) % MOD;
            else ans = (ans + term) % MOD;
        }
        cout << ans << '\n';
    }
    return 0;
}

File: sections/33_inclusion_exclusion/problems/c_bounded_candies/solution.py
import sys


MOD = 1_000_000_007


def prepare(limit: int) -> tuple[list[int], list[int]]:
    fact = [1] * (limit + 1)
    for i in range(1, limit + 1):
        fact[i] = fact[i - 1] * i % MOD
    invfact = [1] * (limit + 1)
    invfact[limit] = pow(fact[limit], MOD - 2, MOD)
    for i in range(limit, 0, -1):
        invfact[i - 1] = invfact[i] * i % MOD
    return fact, invfact


def comb(n: int, r: int, fact: list[int], invfact: list[int]) -> int:
    if r < 0 or r > n:
        return 0
    return fact[n] * invfact[r] % MOD * invfact[n - r] % MOD


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    q = data[0]
    queries = [(data[i], data[i + 1], data[i + 2]) for i in range(1, 3 * q + 1, 3)]
    limit = max(n + s for n, s, _ in queries)
    fact, invfact = prepare(limit)
    out = []
    for n, s, b in queries:
        ans = 0
        step = b + 1
        for bad in range(0, min(n, s // step) + 1):
            remaining = s - bad * step
            term = comb(n, bad, fact, invfact) * comb(remaining + n - 1, n - 1, fact, invfact) % MOD
            ans += term if bad % 2 == 0 else -term
        out.append(str(ans % MOD))
    print("\n".join(out))


if __name__ == "__main__":
    main()
